rust/lance-core/src/container/list.rs

```diff
@@ -0,0 +1,275 @@
+    pub fn push(&mut self, v: T) {
+        match self.inner.back() {
+            Some(last) => {
+                if last.len() == last.capacity() {
+                    let new_cap = if self.limit > 0 && last.capacity() * 2 >= self.limit as usize {
+                        self.limit as usize
+                    } else {
+                        last.capacity() * 2
+                    };
+                    self.inner.push_back(Vec::with_capacity(new_cap));
+                }
+            }
+            None => {
+                self.inner.push_back(Vec::with_capacity(1));
+            }
+        }
+        self.do_push(v);
+    }
+
+    fn do_push(&mut self, v: T) {
+        self.inner.back_mut().unwrap().push(v);
+        self.len += 1;
+    }
+
+    /// Removes the last element from the list.
+    pub fn pop(&mut self) -> Option<T> {
+        match self.inner.back_mut() {
+            Some(last) => {
+                if last.is_empty() {
+                    self.inner.pop_back();
+                    self.pop()
+                } else {
+                    self.len -= 1;
+                    last.pop()
+                }
+            }
+            None => None,
+        }
+    }
+
+    /// Clears the list, removing all elements.
+    /// This will free the memory used by the list.
+    pub fn clear(&mut self) {
+        self.inner.clear();
+        self.len = 0;
+    }
+
+    /// Returns the number of elements in the list.
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Returns whether the list is empty.
+    pub fn is_empty(&self) -> bool {
+        self.inner.is_empty()
+    }
+
+    /// Returns the size of list, including the size of the elements and the
+    /// size of the list itself, and the unused space.
+    /// The element size is calculated using `std::mem::size_of::<T>()`,
+    /// so it is not accurate for all types.
+    /// For example, for `String`, it will return the size of the pointer,
+    /// not the size of the string itself. For that you need to use `DeepSizeOf`.
+    pub fn size(&self) -> usize {
+        let unused_space = match self.inner.back() {
+            Some(last) => last.capacity() - last.len(),
+            None => 0,
+        };
+        (self.len() + unused_space) * std::mem::size_of::<T>()
+            + std::mem::size_of::<Self>()
+            + self.inner.len() * std::mem::size_of::<Vec<T>>()
+    }
+
+    /// Returns an iterator over the elements in the list.
+    pub fn iter(&self) -> ExpLinkedListIter<'_, T> {
+        ExpLinkedListIter::new(self)
+    }
+}
+
+impl<T: DeepSizeOf> DeepSizeOf for ExpLinkedList<T> {
+    fn deep_size_of_children(&self, context: &mut deepsize::Context) -> usize {
+        self.inner
+            .iter()
+            .map(|v| v.deep_size_of_children(context))
+            .sum()
+    }
+}
+
+impl<T> FromIterator<T> for ExpLinkedList<T> {
+    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
+        let mut list = Self::new();
```

Comment:
It might be nice to have a `with_capacity` in the future.  This way if the iterator has a trusted length you can avoid some of the initial allocations.